src/libexpr/flake/call-flake.nix

```
@@ -12,7 +12,36 @@ overrides:
 
 let
 
-  lockFile = builtins.fromJSON lockFileStr;
+  origLockFile = builtins.fromJSON lockFileStr;
+
+  # Build an optimized lock file by consolidating equivalent inputs.
+  lockFile =
+    let
+      nodeCacheKey = key: node:
+        let taint = if overrides ? ${key} then { taintedAs = key; } else {};
+        in builtins.toJSON (node.locked or {} // taint);
+
+      inputsCache = builtins.listToAttrs (builtins.attrValues
+        (builtins.mapAttrs
+          (key: node: { name = nodeCacheKey key node; value = key; })
+          origLockFile.nodes
+        )
+      );
+
+      resolveInputsFromCache = inputs:
+        builtins.mapAttrs
+          (inputName: inputSpec:
+            let resolvedInputName = resolveInput inputSpec;
+            in inputsCache.${nodeCacheKey resolvedInputName lockFile.nodes.${resolvedInputName}})
+          inputs;
+
+      nodes = builtins.mapAttrs
+        (key: node: node // {
+          inputs = resolveInputsFromCache (node.inputs or {});
```

```suggestion
          inputs = builtins.mapAttrs (_nameForArg: deduplicateInputName) (node.inputs or {});
```

- `_` for unused
- `nameForArg` to hint that it's local to that flake, by referring to the only function, `outputs`. I feel like anything more specific would grab too much attention. `local` would have been nice if it weren't ambiguous.